File: apps/Camera-Vitals/Android/app/src/main/java/ai/zetic/demo/cameravitals/signal/HeartRateEstimator.kt
package ai.zetic.demo.cameravitals.signal

import ai.zetic.demo.cameravitals.AppConfig
import ai.zetic.demo.cameravitals.util.clamped

/**
 * Turns a model rPPG waveform (DiffNormalized derivative) into a heart rate.
 * Mirrors rPPG-Toolbox postprocess: cumsum → detrend → bandpass → FFT peak.
 */
object HeartRateEstimator {
    class Result(val bpm: Double, val quality: Double, val filtered: FloatArray)

    fun estimate(waveform: FloatArray, fs: Double): Result? {
        if (waveform.size < 60) return null

        // Integrate (model predicts the derivative of the pulse).
        val cumulative = FloatArray(waveform.size)
        var acc = 0f
        for (i in waveform.indices) { acc += waveform[i]; cumulative[i] = acc }

        val detrended = Detrend.linear(cumulative)
        val filtered = BandpassHR.filtfilt(detrended)

        val peak = Fft.dominantBPM(filtered, fs, AppConfig.HR_BAND_LOW, AppConfig.HR_BAND_HIGH)
            ?: return null

        if (peak.bpm < AppConfig.MIN_BPM || peak.bpm > AppConfig.MAX_BPM) {
            return Result(peak.bpm, 0.0, filtered)  // physiologically implausible
        }

        // Map harmonic SNR (dB) → 0..1 quality (calibrated: clean pulse ≥0.6, noise red).
        val quality = ((peak.snr + 4.0) / 12.0).clamped(0.0, 1.0)
        return Result(peak.bpm, quality, filtered)
    }
}

File: apps/Camera-Vitals/iOS/CameraVitals/Signal/HeartRateEstimator.swift
import Foundation

/// Turns a model rPPG waveform (DiffNormalized derivative) into a heart rate.
/// Mirrors rPPG-Toolbox postprocess: cumsum → detrend → bandpass → FFT peak.
enum HeartRateEstimator {
    struct Result {
        let bpm: Double
        let quality: Double      // 0...1
        let filtered: [Float]
    }

    static func estimate(_ waveform: [Float], fs: Double) -> Result? {
        guard waveform.count >= 60 else { return nil }

        // Integrate (model predicts the derivative of the pulse).
        var cumulative = [Float](repeating: 0, count: waveform.count)
        var acc: Float = 0
        for i in 0..<waveform.count { acc += waveform[i]; cumulative[i] = acc }

        let detrended = Detrend.linear(cumulative)
        let filtered = BandpassHR.filtfilt(detrended)

        guard let peak = FFTAnalyzer.dominantBPM(
            filtered, fs: fs, lo: AppConfig.hrBandLow, hi: AppConfig.hrBandHigh
        ) else { return nil }

        guard peak.bpm >= AppConfig.minBPM, peak.bpm <= AppConfig.maxBPM else {
            return Result(bpm: peak.bpm, quality: 0, filtered: filtered)  // physiologically implausible
        }

        // Map harmonic SNR (dB) → 0...1 quality. Calibrated against scripts: a real pulse
        // (≳ +3 dB) reaches green (≥0.6), a usable pulse amber (≥0.3 ≈ −0.4 dB), pure noise
        // (≈ −1 dB) stays red.
        let quality = ((peak.snr + 4.0) / 12.0).clamped(to: 0...1)
        return Result(bpm: peak.bpm, quality: quality, filtered: filtered)
    }
}
